### PR TITLE
SO-5591: Refactor description acceptability updater request

### PR DESCRIPTION
**Pull request by AAAlinaaa**

### CODE REVIEW

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java [hunk 1]
```diff
@@ -59,49 +63,95 @@ protected boolean canUpdate(TransactionContext context) {
...
+				if (acceptabilityMatches.isEmpty()) {
+					// Case 1.1: No match for acceptability; use an existing member preferably an extension one
+					existingMember = select(refsetMatches, moduleId, Concepts.MODULE_SCT_CORE);
+					final SnomedRefSetMemberIndexEntry.Builder updatedMember = SnomedRefSetMemberIndexEntry.builder(existingMember);
+					final SnomedRefSetMemberIndexEntry oldRevision = updatedMember.build();
+					
+					// Change acceptability, set status to active if required, place it in the supplied module
+					updatedMember.field(SnomedRf2Headers.FIELD_ACCEPTABILITY_ID, newAcceptability.getConceptId());
+					ensureMemberActive(context, existingMember, updatedMember);
+					updateModule(context, existingMember, updatedMember, moduleIdSupplier.apply(componentToUpdate));
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java]: As `moduleIdSupplier.apply(componentToUpdate)` is already extracted to a local variable above, we can make this line shorter:

```suggestion
					updateModule(context, existingMember, updatedMember, moduleId);
```

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java [hunk 2]
```diff
@@ -59,49 +63,95 @@ protected boolean canUpdate(TransactionContext context) {
...
-				updatedMember.field(SnomedRf2Headers.FIELD_ACCEPTABILITY_ID, newAcceptability.getConceptId());
-				ensureMemberActive(context, existingMember, updatedMember);
-				updateModule(context, existingMember, updatedMember, moduleIdSupplier.apply(componentToUpdate));
-				unsetEffectiveTime(existingMember, updatedMember);
-				context.update(oldRevision, updatedMember.build());
-				newLanguageMembersToCreate.remove(languageReferenceSetId);
+				} else {
+					// Case 1.2: There is at least one match for acceptability
+					final List<SnomedReferenceSetMember> activeMatches = acceptabilityMatches.stream()
+							.filter(m -> Objects.equal(newAcceptability.getConceptId(), m.getProperties().get(SnomedRf2Headers.FIELD_ACCEPTABILITY_ID)))
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java]: This filter should look for active members.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java [hunk 3]
```diff
@@ -59,49 +63,95 @@ protected boolean canUpdate(TransactionContext context) {
...
+							context.update(oldRevision, updatedMember.build());
+						}
+					}
+				}
+				
 			} else {
-				// Not acceptable in this language reference set, remove or inactivate if already released
-				if (removeOrDeactivate(context, existingMember, updatedMember)) {
+				// Case 2: No match exists for language reference set
+				newLanguageMembersToCreate.put(languageReferenceSet, newAcceptability);
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java:145]: `newLanguageMembersToCreate` are now collected but don't seem to be used... :perplexed:

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java [hunk 4]
```diff
@@ -117,4 +167,14 @@ protected void doExecute(TransactionContext context, SnomedComponentDocument com
...
 	}
 	
+	private SnomedReferenceSetMember select(List<SnomedReferenceSetMember> members, final String modulePreference, final String secondaryModulePreference) {
+		return members.stream()
+			.filter(member -> member.getModuleId().equals(modulePreference))
+			.findFirst()
+			.orElse(members.stream()
+				.filter(member -> member.getModuleId().equals(secondaryModulePreference))
+				.findFirst()
+				.orElse(members.get(0)));
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java:179]: From the three spots this method is called from, it is OK to use `members.get(0)` as the last fallback option as `members` is non-empty in all cases (first one is "if refset matches non-empty", second is the `else` branch of "if acceptability matches are empty", third is another `else` branch of "if active matches are empty").

To ensure that it stays that way, please add a precondition check to this method.

**Review by apeteri — CHANGES_REQUESTED** [review]
General feedback:

- The class needs a copyright date nudge
- There is one failed integration test case: https://github.com/b2ihealthcare/snow-owl/actions/runs/4106090555/jobs/7083866020

```
Failures: 
  SnomedDescriptionApiTest.updateAcceptabilityWithMemberReactivation:561->lambda$3:561 Reference set member should be active expected:<true> but was:<false>

Tests run: 767, Failures: 1, Errors: 0, Skipped: 7
```

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java (hunk 3 above)
Comment (AAAlinaaa) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java:145]: They have always been collected, and are used.

Comment (apeteri) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java:145]: Oops, missed the unchanged block of code at the end of the file, sorry about that.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java (hunk 2 above)
Comment (AAAlinaaa) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedDescriptionAcceptabilityUpdateRequest.java]: SnomedDescriptionApiTest no longer fails after this fix.

**Review by apeteri — APPROVED** [review]
LGTM!

**Review by nagyo — APPROVED** [review]
Overall looks good to me, but given this is a core SNOMED functionality, I would love to see some unit tests covering those "Case 1" / "Case 1.1" ... etc scenarios.

**Comment by nagyo** [comment]
And I suppose a 8.x counterpart should be created as well.

**Comment by apeteri** [comment]
Added Gábor's suggestions as a JIRA ticket.

**Pull request merged**